android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -39,8 +41,52 @@ class TranslationsMiddleware(
     ) {
         // Pre process actions
         when (action) {
+            // Initially adding the tab
+            is TabListAction.AddTabAction -> {
+                if (translationsState(context, action.tab.id)?.isEngineSupported == null) {
+                    context.store.dispatch(
+                        TranslationsAction.OperationRequestedAction(
+                            tabId = action.tab.id,
+                            operation = TranslationOperation.FETCH_IS_ENGINE_SUPPORTED,
+                        ),
+                    )
+                }
+            }
+
+            // Restoring a tab from a closed app
+            is TabListAction.RestoreAction -> {
+                if (translationsState(context, action.selectedTabId)?.isEngineSupported == null) {
+                    action.selectedTabId?.let { id ->
+                        context.store.dispatch(
+                            TranslationsAction.OperationRequestedAction(
+                                tabId = id,
+                                operation = TranslationOperation.FETCH_IS_ENGINE_SUPPORTED,
+                            ),
+                        )
+                    }
+                }
+            }
+
+            // Switching to a new tab
+            is TabListAction.SelectTabAction -> {
+                if (translationsState(context, action.tabId)?.isEngineSupported == null) {
+                    context.store.dispatch(
+                        TranslationsAction.OperationRequestedAction(
+                            tabId = action.tabId,
+                            operation = TranslationOperation.FETCH_IS_ENGINE_SUPPORTED,
+                        ),
+                    )
+                }
+            }
+
             is TranslationsAction.OperationRequestedAction -> {
                 when (action.operation) {
+                    TranslationOperation.FETCH_IS_ENGINE_SUPPORTED -> {
+                        scope.launch {
+                            requestEngineSupport(context, action.tabId)
+                        }
+                    }
```

Comment:
I'm only seeing that this operation is only really dispatched from this middleware, and only really used in this middleware outside of resetting the error state. I'm wondering if we need the additional operation/action dispatch, or if we can just call `requestEngineSupport` directly from things like `AddTabAction`. 

If the `isEngineSupported` is global state as per the questions above, could we even maybe use `BrowserAction.Init`? I do seem to remember that you looked into that but can't recall the details

Reply:
Thanks, updated the pattern! 

I can use `BrowserAction.Init` now with the tab independent state. The earlier issue was I needed a tabId to set on, if I set per tab. I still might have to figure out alternatives for tab initialization, but `BrowserAction.Init` seems much more stable for this patch.